### ISSUE TITLE
Use interrupts for Wakers

### ISSUE THREAD

**Issue by lights0123** [lights0123/async-avr#3]
This is currently blocked waiting for [an LLVM bug](https://gitter.im/avr-rust/Lobby?at=5f1ca63fbc41f3681724bbba) to be fixed, but we can work on other things to get ready. [They tried to fix it](https://reviews.llvm.org/D82536), but there were some problems with the patch.

* [ ] Wait for LLVM to fix [their bug](https://gitter.im/avr-rust/Lobby?at=5f1ca63fbc41f3681724bbba)
* [ ] Send a PR to [Rahix/avr-hal](https://github.com/Rahix/avr-hal) to allow enabling interrupts
* [ ] Figure out who should be responsible for listening for interrupts
  * Probably the user in the binary crate, but how?
  * Expose APIs that are called when an interrupt occurs?

**Comment by Rahix** [comment]
Maybe, for the time being, one could implement a 'custom waker API' that does not use any function pointers?  E.g. a global singleton "waker object" that is used unconditionally in all contexts?

**Comment by lights0123** [comment]
@Rahix Yeah, that's definitely possible. However, it will still introduce problems with third-party crates—things from the `futures` crate that work without `alloc`, for example.

**Comment by jeremysalwen** [comment]
Looks like the LLVM bug was fixed! https://reviews.llvm.org/D87631

**Comment by lights0123** [comment]
@jeremysalwen I've taken a break from this (well, because of this issue), but if you'd like to follow along, it looks like someone else is working on an executor based on my work: https://www.reddit.com/r/rust/comments/rxmy9k/comment/hrkb7il/?utm_source=share&utm_medium=web2x&context=3

**Comment by mfp20** [comment]
Any update on this? Triggering on IRQ is possible ( https://github.com/matthunz/async-hal ). I'm at my first project in Rust and I'm banging my head on your code to figure out how to use irqs instead of block_on(); well, "instead of" a blocking executor.
The project I linked above looks a bit too heavy for AVR; it would be much better to mod your code instead.

**Comment by lights0123** [comment]
embassy is the leading async framework for embedded, so I'd take a look at ThreadPender from a sample integration with it: https://github.com/niclashoyer/embassy/blob/avr/embassy-executor/src/arch/avr.rsIt's checked in the hot loop at the end. An interrupt can change that atomic variable to wake up the loop.On Aug 23, 2024, at 9:22 AM, mfp20 ***@***.***> wrote:﻿
Any update on this? Triggering on IRQ is possible ( https://github.com/matthunz/async-hal ). I'm at my first project in Rust and I'm banging my head on your code to figure out how to use irqs instead of block_on(); well, "instead of" a blocking executor.
The project I linked above looks a bit too heavy for AVR; it would be much better to mod your code instead.

**Comment by mfp20** [comment]
from [embassy](https://github.com/niclashoyer/embassy/blob/avr/embassy-executor/src/arch/avr.rs):

```
#[cfg(feature = "executor-interrupt")]
compile_error!("`executor-interrupt` is not supported with `arch-avr`.");
```

And in general, avr port isn't great in embassy.
